### ISSUE TITLE
Dubbo filter listener onResponse method NPE

### ISSUE THREAD

**Issue by zmapleshine** [alibaba/spring-cloud-alibaba#1494]
我们鼓励使用英文，如果不能直接使用，可以使用翻译软件，您仍旧可以保留中文原文。
We recommend using English. If you are non-native English speaker, you can use the translation software. 

**Which Component**
Dubbo 2.7.6

**Describe what problem you have encountered**

I implemented metrics records in the filter, and there was no problem in the local environment test. In the online environment, I had an NPE error in the onResponse (implemented the filter listening interface) callback method. The specific code is as follows:

```java
@Slf4j
@Activate(group = {CommonConstants.PROVIDER, CommonConstants.CONSUMER}, order = 2)
public class DubboMetricsFilter implements Filter, Filter.Listener {
  private static final Map<Long, Recorder> RECORDER_MAP = new ConcurrentHashMap<>();
  private static final AtomicLong REQ_ID = new AtomicLong(0);
  //Omit others...

  @Override
    public Result invoke(Invoker<?> invoker, Invocation invocation) {
        Long reqId = REQ_ID.getAndIncrement();
        invocation.setObjectAttachment(“reqId”, reqId);
        Recorder recorder = new Recorder(invoker,invocation);
        recorder.recordStart();
        return invoker.invoke(invocation);
    }


@Override
    public void onResponse(Result appResponse, Invoker<?> invoker, Invocation invocation) {
        Long reqId = (Long) invocation.getObjectAttachment(REQ_ID_KEY);
        Recorder recorder = RECORDER_MAP.get(reqId);
        recorder.recordSuccess();//  be possible to occur NPE cause recorder is null
        recorder.recordFinished();
        RECORDER_MAP.remove(reqId);
    }

    @Override
    public void onError(Throwable t, Invoker<?> invoker, Invocation invocation) {
       
        String result = "error";
        if (t instanceof RpcException) {
            RpcException e = (RpcException) t;
            result = "others";
        }
        Long reqId = (Long) invocation.getObjectAttachment(REQ_ID_KEY);
        Recorder recorder = RECORDER_MAP.get(reqId); 
        recorder.recordError(Collections.singletonMap("result", result));//  be possible to occur NPE cause recorder is null
        recorder.recordFinished();
        RECORDER_MAP.remove(reqId);
    }
```

```java
Recorder recorder = RECORDER_MAP.get(reqId);  // recorder null
recorder.xxxx ///NullPointException!!!!
```

Is it possible that the onResponse callback method will be executed multiple times in a concurrent environment but the invoke method will only be executed once?

In other words, the Attachment value in the invocation of the client onResponse callback may be modified by the provider？

In addition, the possibility that the object corresponding to the reqId of RECORD_MAP is removed cannot be expected.

**Describe what information you have read**
eg. I have read the reference doc of Sentinel
